### PR TITLE
Add step.id 

### LINKED ISSUE #96: Add unique IDs to every step
Add a unique ID to every step in a mission. The ID can then be used to manage inspections within isar-robot such that ISAR itself only has to manage the step ID.

This will remove the need for `vendor_mission_id` where this is present.

### PR DESCRIPTION
**Pull request by autt**
Resolves #96 and #149

This PR changes the robot interface, and will require corresponding change in ISAR robot packages.

There are PRs for [isar-robot](https://github.com/equinor/isar-robot/pull/18) and [isar-turtlebot](https://github.com/equinor/isar-turtlebot/pull/31) ready for review.

### CODE REVIEW

File: src/isar/state_machine/states/monitor.py [hunk 1]
```diff
@@ -91,14 +89,13 @@ def _run(self):
         self.state_machine.to_next_state(next_state)
 
     def _mission_finished(
-        self, mission_status: MissionStatus, instance_id: int
+        self,
+        mission_status: MissionStatus,
     ) -> bool:
```
Comment (tsundvoll) [inline src/isar/state_machine/states/monitor.py:94]: Function definition could go on one line

File: src/robot_interface/models/mission/task.py [hunk 2]
```diff
@@ -68,7 +69,8 @@ class DriveToPose(MotionTask):
     """
 
     pose: Pose
-    task_name: Literal["drive_to_pose"] = "drive_to_pose"
+    name: Literal["drive_to_pose"] = "drive_to_pose"
+    id: UUID = field(default_factory=uuid4)
```
Comment (tsundvoll) [inline src/robot_interface/models/mission/task.py:73]: Could the id attribute be a part of the Task dataclass, since all subclasses needs it?

File: src/robot_interface/robot_interface.py [hunk 3]
```diff
@@ -45,12 +45,12 @@ def mission_scheduled(self) -> bool:
...
     def mission_status(self, mission_id: Any) -> MissionStatus:
-        """Retrieves status of the current executing mission for the robot.
+        """Retrieves status of the current executing mission task for the robot.
 
         Parameters
         ----------
-        mission_id : Any
-            Unique identifier for the mission which the status should be checked for
+        task_id : UUID
+            Unique identifier for the task which the status should be checked for
```
Comment (tsundvoll) [inline src/robot_interface/robot_interface.py:53]: Docstring is renamed, but the parameter in the function def is still called mission_id

File: tests/isar/services/readers/test_mission_reader.py [hunk 4]
```diff
@@ -94,6 +94,6 @@ def test_thermal_image_task(mission_reader):
 
     assert isinstance(task, TakeThermalImage)
     assert hasattr(task, "target")
-    assert task.task_name == "take_thermal_image"
+    assert task.name == "take_thermal_image"
```
Comment (tsundvoll) [inline tests/isar/services/readers/test_mission_reader.py:97]: Add check for task.id
```suggestion
    assert task.name == "take_thermal_image"
    assert hasattr(task, "id")
```

File: tests/mocks/status.py [hunk 5]
```diff
@@ -17,7 +16,6 @@ def mock_status(
     scheduled_status = Status(
         mission_status=mission_status,
         mission_in_progress=mission_in_progress,
-        current_mission_instance_id=current_mission_instance_id,
```
Comment (tsundvoll) [inline tests/mocks/status.py:20]: We should make sure that the mock contains an ID on the step. Is this done? If not, and the tests are not failing, we probably should extend the test to cover this.

File: src/isar/mission_planner/local_planner.py [hunk 6]
```diff
@@ -56,7 +56,7 @@ def read_mission_from_file(mission_path: Path) -> Optional[Mission]:
             dataclass_dict=mission_dict,
             target_dataclass=Mission,
             cast_config=[Frame],
-            strict_config=True,
+            strict_config=False,
```
Comment (tsundvoll) [inline src/isar/mission_planner/local_planner.py:59]: Why is this changed?

**Review by tsundvoll — CHANGES_REQUESTED** [review]
Great work, this simplifies a lot! A few suggestions in the comments.

File: src/robot_interface/robot_interface.py [hunk 7]
```diff
@@ -92,15 +88,11 @@ def log_status(
         raise NotImplementedError
 
     @abstractmethod
-    def get_inspection_references(
-        self, vendor_mission_id: Any, current_task: Task
-    ) -> Sequence[Inspection]:
+    def get_inspection_references(self, current_task: Task) -> Sequence[Inspection]:
```
Comment (tsundvoll) [inline src/robot_interface/robot_interface.py:91]: Does it have to be the current task we want to get the inspection reference from? Could this instead be renamed to just task? This is maybe a separate PR; but still an interesting thought.
```suggestion
    def get_inspection_references(self, task: Task) -> Sequence[Inspection]:
```

File: src/isar/models/communication/status.py [hunk 8]
```diff
@@ -10,7 +10,6 @@
 class Status:
     mission_status: Optional[MissionStatus]
     mission_in_progress: bool
-    current_mission_instance_id: Optional[int]
     current_mission_task: Optional[Task]
```
Comment (aestene) [inline src/isar/models/communication/status.py:13]: Can we rename this to simply `current_task` while we're at it? The concept of `current_task` should be unique in this context now.

File: src/isar/state_machine/states/monitor.py [hunk 9]
```diff
@@ -91,14 +89,13 @@ def _run(self):
...
         if mission_status == MissionStatus.Unexpected:
-            self.logger.error(
-                f"Mission status on task {instance_id} returned and unexpected status string"
-            )
+            self.logger.error("Mission status returned an unexpected status string")
         elif mission_status == MissionStatus.Failed:
-            self.logger.warning(f"Mission instance {instance_id} failed...")
+            self.logger.warning("Mission failed...")
```
Comment (aestene) [inline src/isar/state_machine/states/monitor.py:98]: Strictly speaking the Mission didn't fail, just a Task within the mission so this log message is not correct.

Would recommend logging something like "Task failed. Task ID: id" or something along those lines.

Comment (aestene) [reply src/isar/state_machine/states/monitor.py:98]: We discussed renaming `MissionStatus` to `TaskStatus` as it's the status of a single `Task` and not the mission as a whole. If you wish to include that here you may, otherwise I can create a new issue to do so.

File: src/robot_interface/models/mission/task.py (hunk 2 above)
Comment (aestene) [reply src/robot_interface/models/mission/task.py:73]: This would be preferrable, however dataclasses in python have a weird limitation here.

If you define a field with a default value in a parent dataclass the inheriting dataclass cannot have any fields without default values. This is because the inheritance for dataclasses "stacks" fields in such a way that the default field then comes before the non-default field.

Comment (aestene) [reply src/robot_interface/models/mission/task.py:73]: I think it would be solved by using regular classes where the behavior might be a bit different if I remember correctly. Although that's not really a solution.

The same limitation is why `tag_id` is present in both `TakeImage` and `TakeThermalImage` but not in `InspectionTask`.

File: tests/mocks/status.py (hunk 5 above)
Comment (autt) [reply tests/mocks/status.py:20]: There is not an ID field in Status anymore. The id can be found in current_task.id.

File: src/robot_interface/robot_interface.py (hunk 7 above)
Comment (aestene) [reply src/robot_interface/robot_interface.py:91]: Good idea. It would change the requirements to the implementation of the robot though depending on what the robot does.

For example, would we expect to retrieve inspections for "any" task id? I would assume just for the task IDs in the current mission.

File: src/isar/mission_planner/local_planner.py (hunk 6 above)
Comment (autt) [reply src/isar/mission_planner/local_planner.py]: Reverted

File: src/robot_interface/models/mission/task.py (hunk 2 above)
Comment (autt) [reply src/robot_interface/models/mission/task.py:73]: I agree with @aestene here. I suggest that we keep this implementation for now, but looking at ways to improve the dataclass implmentation is left for issue #133.

File: src/isar/state_machine/states/monitor.py (hunk 9 above)
Comment (autt) [reply src/isar/state_machine/states/monitor.py]: In general, the term `mission` is not used consistently and does not have a unique meaning in ISAR. Renaming is in process and will be added to this PR. `MissionStatus ` -> `TaskStatus` is among those changes.

File: src/robot_interface/robot_interface.py (hunk 7 above)
Comment (autt) [reply src/robot_interface/robot_interface.py:92]: Interesting idea. It is now left for #147

**Review by aestene — APPROVED** [review]
LGTM. 
I would await merging until the implementation for all robot repos are ready in this case.

File: src/robot_interface/models/mission/task.py [hunk 10]
```diff
@@ -79,6 +81,7 @@ class DockingProcedure(MotionTask):
 
     behavior: Literal["dock, undock"]
     name: Literal["docking_procedure"] = "docking_procedure"
+    id: UUID = field(default_factory=uuid4)
```
Comment (vetlek) [inline src/robot_interface/models/mission/task.py:84]: Instead of setting `id` for all step classes, you can add `id: UUID = field(default_factory=uuid4, init=false)` in the base class

**Review by vetlek — CHANGES_REQUESTED** [review]

File: src/robot_interface/models/mission/task.py (hunk 10 above)
Comment (autt) [reply src/robot_interface/models/mission/task.py]: Good catch :+1:

**Review by vetlek — APPROVED** [review]
LGTM

**Review by tsundvoll — APPROVED** [review]

**Review by aestene — APPROVED** [review]
